libr/util/axml.c

```diff
@@ -0,0 +1,430 @@
+			return NULL;
+		}
+
+		ut16 *start16 = (ut16 *)start;
+
+		// If >0x7fff, stored as a big-endian ut32
+		ut32 n = r_read_le16 (start16++);
+		if (n & 0x8000) {
+			n |= ((n & 0x7fff) << 16) | r_read_le16 (start16++);
+		}
+
+		// Size of UTF-16LE without NULL
+		n *= 2;
+
+		name = calloc (n * 2 + 1, 1);
+
+		if ((ut64)start16 > (ut64)data + data_size - sizeof(ut32) - n - 1) {
+			free (name);
+			return NULL;
+		}
+
+		// If UTF-16LE, decode to UTF-8 so we can print it to the screen
+		if (r_str_utf16_to_utf8 ((ut8 *)name, n * 2, (const ut8 *)start16, n, true) < 0) {
+			free (name);
+			eprintf ("Failed to decode UTF16-LE\n");
+			return NULL;
+		}
+
+		if (length) {
+			*length = n;
+		}
+	}
+
+	return name;
+}
+
+static char *resource_value(string_pool_t *pool, const ut8 *data, ut64 data_size,
+		resource_value_t *value) {
+	switch (value->type) {
+	case RESOURCE_NULL:
+		return r_str_new ("");
+	case RESOURCE_REFERENCE:
+		return r_str_newf ("@0x%x", value->data.d);
+	case RESOURCE_STRING:
+		return string_lookup (pool, data, data_size, r_read_le32 (&value->data.d), NULL);
+	case RESOURCE_FLOAT:
+		return r_str_newf ("%f", value->data.f);
+	case RESOURCE_INT_DEC:
+		return r_str_newf ("%d", value->data.d);
+	case RESOURCE_INT_HEX:
+		return r_str_newf ("0x%x", value->data.d);
+	case RESOURCE_BOOL:
+		return r_str_newf (value->data.d ? "true" : "false");
+	default:
+		eprintf ("Resource type is not recognized: %#x\n", value->type);
+		break;
+	}
+	return r_str_new ("null");
+}
+
+ut32 depth = 0;
+
+static bool dump_element(RStrBuf *sb, string_pool_t *pool, namespace_t *namespace,
+		const ut8 *data, ut64 data_size, start_element_t *element, 
+		const ut32 *resource_map, ut32 resource_map_length, bool start) {
+	ut32 i;
+
+	char *name = string_lookup (pool, data, data_size, r_read_le32 (&element->name), NULL);
+	for (i = 0; i < depth; i++) {
+		r_strbuf_appendf (sb, "\t");
+	}
+
+	if (start) {
+		r_strbuf_appendf (sb, "<%s", name);
+		ut16 count = r_read_le16 (&element->attribute_count);
+		if (depth == 0 && namespace) {
+			char *key = string_lookup (pool, data, data_size, namespace->prefix, NULL);
+			char *value = string_lookup (pool, data, data_size, namespace->uri, NULL);
+			r_strbuf_appendf (sb, " xmlns:%s=\"%s\"", key, value);
+			free(key);
+			free(value);
+		}
+
+		if (count != 0) {
+			r_strbuf_appendf (sb, " ");
+		}
+
+		for (i = 0; i < count; i++) {
+			attribute_t a = element->attributes[i];
+			ut32 key_index = r_read_le32 (&a.name);
+			const char *key = (const char *)string_lookup (pool, data, data_size, key_index, NULL);
+			bool resource_key = false;
+			// If the key is empty, it is a cached resource name
+			if (key && *key == '\0') {
+				free ((char *)key);
+				resource_key = true;
+				if (resource_map && key_index < resource_map_length) {
+					ut32 resource = resource_map[key_index];
+					resource -= 0x1010000;
+					if (resource >= 0 && resource < ANDROID_ATTRIBUTE_NAMES_SIZE) {
```

Comment:
Better check against 0x1010000 before substraction. Ans avoid the unsigned comparison to 0 below